src/main/java/uk/gov/hmcts/reform/slc/services/steps/getpdf/PdfCreator.java

```diff
@@ -1,33 +1,109 @@
 package uk.gov.hmcts.reform.slc.services.steps.getpdf;
 
 import org.apache.http.util.Asserts;
+import org.springframework.beans.factory.annotation.Autowired;
 import org.springframework.stereotype.Service;
 import uk.gov.hmcts.reform.pdf.service.client.PDFServiceClient;
+import uk.gov.hmcts.reform.pdf.service.client.exception.PDFServiceClientException;
+import uk.gov.hmcts.reform.slc.logging.AppInsights;
+import uk.gov.hmcts.reform.slc.model.Document;
 import uk.gov.hmcts.reform.slc.model.Letter;
 
+import java.time.Duration;
+import java.time.Instant;
 import java.util.List;
 
 import static java.util.stream.Collectors.toList;
 
 @Service
 public class PdfCreator {
 
+    @Autowired
+    private AppInsights insights;
+
     private final PDFServiceClient client;
 
+    private Letter letter;
+
     public PdfCreator(PDFServiceClient client) {
         this.client = client;
     }
 
     public List<PdfDoc> create(Letter letter) {
         Asserts.notNull(letter, "letter");
+        this.letter = letter;
+
+        List<GeneratorResult> results = letter.documents
+            .stream()
+            .map(this::generatePdf)
+            .peek(this::trackGeneratorResult)
+            .collect(toList());
 
-        return letter.documents
+        return results
             .stream()
-            .map(d -> client.generateFromHtml(d.template.getBytes(), d.values))
-            .map(content -> new PdfDoc(
-                    FileNameGenerator.generateFor(letter.type, letter.service, content, "pdf"),
-                    content
-                )
-            ).collect(toList());
+            .peek(this::checkForException)
+            .map(this::getPdfDoc)
+            .collect(toList());
+    }
+
+    private GeneratorResult generatePdf(Document document) {
+        GeneratorResult result = new GeneratorResult();
+        Instant start = Instant.now();
+
+        try {
+            byte[] pdf = client.generateFromHtml(document.template.getBytes(), document.values);
+
+            result.successful(Duration.between(start, Instant.now()), pdf);
+        } catch (PDFServiceClientException exception) {
+            result.failed(Duration.between(start, Instant.now()), exception);
+        }
+
+        return result;
```

Comment:
why would we keep going if first pdf generation fails??

Reply:
last commit can be extracted to different pr. was working on some tracking solution. can be completely obliterated